[exporter/prometheusremotewrite] deprecate export_created_metric

pokgak opened the pull request:
#### Description

Deprecate export_created_metric config option

#### Link to tracking issue
Fixes #35003

pokgak committed:
[exporter/prometheusremotewrite] deprecate export_created_metric

ArthurSens commented on exporter/prometheusremotewriteexporter/config.go:
```diff
@@ -45,6 +45,7 @@ type Config struct {
 	TargetInfo *TargetInfo `mapstructure:"target_info,omitempty"`
 
 	// CreatedMetric allows customizing creation of _created metrics
+	// Deprecated: This feature is useless in PRWv1 and and PRWv2 already have the created timestamp in the proto
```
```suggestion
	// Deprecated: This feature is useless in PRWv1 and and PRWv2 already have the created timestamp in the proto
```
```suggestion
	// Deprecated[0.113.0]: The feature doesn't provide the expected behavior. Use Prometheus remote-write v2 to enable sending Created Timestamps.
	// This feature is planned to be removed in v0.115.0
```

ArthurSens commented on exporter/prometheusremotewriteexporter/README.md:
```diff
@@ -59,7 +59,7 @@ The following settings can be optionally configured:
 - `target_info`: customize `target_info` metric
   - `enabled` (default = true): If `enabled` is `true`, a `target_info` metric will be generated for each resource metric (see https://github.com/open-telemetry/opentelemetry-specification/pull/2381).
-- `export_created_metric`:
+- `export_created_metric` (deprecated):
```
```suggestion
- `export_created_metric`: `WARNING` Deprecated and planned for removal in v0.115.0. See [related issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35003) for more information. 
```

ArthurSens reviewed (commented):
Thank you @pokgak!

Not sure if this is documented well, but when deprecating we can put the functionality behind a feature gate that is enabled by default. One example of feature gate can be seen [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5a3892e8d6e802c70ee4366f8194948adbda7ae9/pkg/translator/prometheus/normalize_label.go#L13-L18). In the example though the stability is `StageAlpha` (disabled by default) and we want it to considered `StageBeta` so it's enabled by default.

In version 0.114.0 we'll finally change it to Alpha and in 0.115.0 finally remove the option.

pokgak committed:
Apply suggestions from code review

Co-authored-by: Arthur Silva Sens <arthursens2005@gmail.com>

pokgak committed:
use feature gate to set default value

pokgak commented:
@ArthurSens correct me if I'm wrong but isn't the feature already disabled by default so we should be using `StageAlpha` instead?

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7ba5aa30c1dda62f59ddd6f8a712397f4f9ce55c/exporter/prometheusremotewriteexporter/config.go#L105-L109

I pushed a commit using `StageBeta` but note that this will change the default value of the `export_created_metric.enabled` to `true`.

pokgak committed:
updated feature gate description

ArthurSens commented on exporter/prometheusremotewriteexporter/config.go:
```diff
@@ -104,7 +114,7 @@ func (cfg *Config) Validate() error {
 	if cfg.CreatedMetric == nil {
 		cfg.CreatedMetric = &CreatedMetric{
-			Enabled: false,
+			Enabled: exportCreatedMetricGate.IsEnabled(),
```
```suggestion
			Enabled: false,
```

Let's go back to false!

ArthurSens commented on exporter/prometheusremotewriteexporter/config.go:
```diff
@@ -83,6 +86,13 @@ type RemoteWriteQueue struct {
 var _ component.Config = (*Config)(nil)
 
+var exportCreatedMetricGate = featuregate.GlobalRegistry().MustRegister(
+	"exporter.prometheusremotewriteexporter.exportCreatedMetric",
+	featuregate.StageBeta,
+	featuregate.WithRegisterDescription("Set the default value of the export_created_metric feature."),
+	featuregate.WithRegisterReferenceURL("https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35003"),
```
```suggestion
	"exporter.prometheusremotewriteexporter.deprecateCreatedMetric",
	featuregate.StageBeta,
	featuregate.WithRegisterDescription("Feature gate used to control the deprecation of created metrics."),
	featuregate.WithRegisterReferenceURL("https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35003"),
```

ArthurSens reviewed (commented):
> @ArthurSens correct me if I'm wrong but isn't the feature already disabled by default so we should be using `StageAlpha` instead?
> 
> https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7ba5aa30c1dda62f59ddd6f8a712397f4f9ce55c/exporter/prometheusremotewriteexporter/config.go#L105-L109
> 
> I pushed a commit using `StageBeta` but note that this will change the default value of the `export_created_metric.enabled` to `true`.

Yes you're correct, I believe I wasn't clear with my previous comment, let me re-phrase it.

What we want to do is add yet another check when we're generating the created metric.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3458d516819f31e87a271dc7855529519917dc9c/pkg/translator/prometheusremotewrite/helper.go#L277-L280

Here we'd switch to:
```go
 if settings.ExportCreatedMetric && startTimestamp != 0 && exportCreatedMetricGate.IsEnabled() { 
 	labels := createLabels(baseName+createdSuffix, baseLabels) 
 	c.addTimeSeriesIfNeeded(labels, startTimestamp, pt.Timestamp()) 
 } 
```

And then we'll have to update one test to make sure the feature-gate works as expected:
* [TestPrometheusConverter_AddHistogramDataPoints](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3458d516819f31e87a271dc7855529519917dc9c/pkg/translator/prometheusremotewrite/helper_test.go#L785)

Add a new boolean to the test case struct, and in the the Run you can add:
```go
oldValue := exportCreatedMetricGate.IsEnabled()
testutil.SetFeatureGateForTest(t, exportCreatedMetricGate, tt.isGateEnabled)
defer testutil.SetFeatureGateForTest(t, exportCreatedMetricGate, oldValue)
```

pokgak committed:
Apply suggestions from code review

Co-authored-by: Arthur Silva Sens <arthursens2005@gmail.com>

pokgak committed:
update test case

ArthurSens reviewed (approved):
LGTM, thank you!

Could a maintainer take a look and trigger CI?

ArthurSens commented:
Actually, @pokgak could you rebase your PR on top of main? It looks like you've got some merge conflicts

pokgak commented:
@ArthurSens thanks for the detailed steps.

In case you're reviewing this already, I only followed based on your steps for now. I think I need to also add test case for when the feature gate is false to make sure it's working correctly but haven't had time yet will do it tomorrow.

pokgak committed:
add test case for exportCreatedGate disabled

pokgak commented:
@ArthurSens I added the test case and rebased to main. Also updated another test case [TestPrometheusConverter_AddSummaryDataPoints](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3458d516819f31e87a271dc7855529519917dc9c/pkg/translator/prometheusremotewrite/helper_test.go#L677) since it uses the exportCreatedMetric config option too. Should be ready for review again.

ArthurSens reviewed (approved):
Thanks!

dashpole commented on .chloggen/35003-deprecate-export_created_metric.yaml:
```diff
@@ -0,0 +1,25 @@
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Since the feature gate is beta in this release, we should let people know how they can re-enable the feature temporarily.
```suggestion
subtext: Disable the exporter.prometheusremotewriteexporter.deprecateCreatedMetric feature gate to temporarily re-enable the created metric.
```

ArthurSens replied:
It's the opposite, isn't it? a Beta feature gate is enabled by default, so you don't need to do anything to enable this.

Once we move to alpha then people need to manually re-enable the functionality

ArthurSens replied:
https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate#:~:text=A%20beta%20stage%20where%20the%20feature%20has%20been%20well%20tested%20and%20is%20enabled%20by%20default%20but%20can%20be%20disabled%20through%20a%20Gate.

dashpole replied:
oh!  The feature gate should always progress from alpha -> beta -> stable.  Stable should mean that it is permanently disabled.  Alpha should mean that it is still enabled.

pokgak replied:
No worries :) I'll update the PR accordingly.

pokgak committed:
Update .chloggen/35003-deprecate-export_created_metric.yaml

Co-authored-by: David Ashpole <dashpole@google.com>

pokgak committed:
fix featuregate default to stage alpha

ArthurSens reviewed (approved):
Thank again!

dashpole committed:
Merge branch 'main' into exporter/prometheusremotewriteexporter/deprecate-created-metric

TylerHelmuth commented:
@pokgak please handle merge conflicts

dashpole committed:
Merge branch 'main' into exporter/prometheusremotewriteexporter/deprecate-created-metric

pokgak commented:
thanks @dashpole @TylerHelmuth for the review and merge! also thanks to @ArthurSens for the guidance, if you got any other issues you need help with feel free to ping me.
